pytorch_lightning/strategies/ddp_spawn.py

```diff
@@ -119,10 +121,15 @@ def setup(self, trainer: "pl.Trainer") -> None:
             self.model = self._layer_sync.apply(self.model)
 
         # skip wrapping the model if we are not fitting as no gradients need to be exchanged
-        trainer_fn = self.lightning_module.trainer.state.fn
+        trainer_fn = trainer.state.fn
         if trainer_fn == TrainerFn.FITTING:
             self.configure_ddp()
 
+        # set up optimizers after the wrapped module has been moved to the device
+        self.setup_optimizers(trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
```

Comment:
wondering if `self.setup_optimizers` should include `optimizers_to_device` by default, so users only ever need to call `setup_optimizers`

Reply:
Is `PostLocalSGDState` not supported for DDPSpawn? Why do we have a feature divergence there ?

Reply:
This is already in the parent setup method. I would prefer to rely as much as possible on the parent methods from the base `Strategy`.

Maybe the base class should be:

```py
    def setup(self, trainer: "pl.Trainer") -> None:
        """Setup plugins for the trainer fit and creates optimizers.

        Args:
            trainer: the trainer instance
        """
        self.accelerator.setup(trainer)
        self.model_to_device()
        self.configure_model()
        self.setup_optimizers(trainer)
        self.setup_precision_plugin()
        optimizers_to_device(self.optimizers, self.root_device)
```

Reply:
@awaelchli @carmocca @justusschock

Reply:
as @awaelchli  this doesn't work even now. Bagua expects the optimizers to be initialized before wrapping the model.